iceoryx_examples/waitset/ice_waitset_trigger.cpp

```diff
@@ -0,0 +1,234 @@
+// Copyright (c) 2020 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iceoryx_posh/popo/wait_set.hpp"
+#include "iceoryx_posh/runtime/posh_runtime.hpp"
+#include "iceoryx_utils/cxx/optional.hpp"
+
+#include <iostream>
+#include <thread>
+
+// The two events the MyTriggerClass offers
+enum class MyTriggerClassEvents
+{
+    PERFORMED_ACTION,
+    ACTIVATE
+};
+
+// Triggerable class which has two events an both events can be
+// attached to a WaitSet.
+class MyTriggerClass
+{
+  public:
+    MyTriggerClass() = default;
+    ~MyTriggerClass() = default;
+
+    // IMPORTANT: For now the WaitSet does not support that the origin is moved
```

Comment:
As I see it, the user is responsible for this. One can of course write classes (and derive from them) where this is handled in the copy or move ctor (copy I find dubious though, what about the orginal? why detach it?). This just requires a suitable API of the waitset to modify certain trigger attributes (like the callback here).

I also would be more general and maybe even allow to re-register the whole callback (i.e. including function, not just the object pointer). In the end this feels highly specialized and I am not sure we actually need this.

Reply:
At the moment we absolutely need this since the WaitSet is storing a pointer to this class and a method pointer for the hasTriggeredCallback and for the resetCallback. If the class would be moved or copied we could end up with dangling pointers.

This problem is solvable but for this we have to inform the waitset that we have moved so that it can update the pointer to the object.

This point is already mentioned as todo in the issue: https://github.com/eclipse/iceoryx/issues/341